### ISSUE TITLE
Getting multple executions of handle_info on extra_columns subscription

### ISSUE THREAD

**Issue by camstuart** [cheerfulstoic/ecto_watch#26]
Hello,

Firstly, thank you so much for this amazing library, when I heard it mentioned on "thinking elixir podcast" while I was driving I immediately had to pull over and make a note!

I am using the `extra_columns` feature in an effort to have my live view process update when a message record arrives/updates for a given `organisation_id`.

In `application.ex`: (commented out insert and delete to try an isolate the issue)

```elixir
{EctoWatch,
 repo: Relay.Repo,
 pub_sub: Relay.PubSub,
 watchers: [
   # {Message, :inserted, extra_columns: [:organisation_id]},
   # {Message, :deleted, extra_columns: [:organisation_id]},
   {Message, :updated, extra_columns: [:organisation_id]}
 ]}
```

The `mount` function in the live view (@org_id, @user_id are hard coded)

```elixir
@impl true
def mount(params, _session, socket) do
  if connected?(socket) do
    # EctoWatch.subscribe({Message, :inserted})
    # EctoWatch.subscribe({Message, :deleted})
   EctoWatch.subscribe({Message, :updated}, {:organisation_id, @org_id})
  end

  messages = Messaging.query_messages_by_user(@org_id, @user_id) |> Repo.all()

  socket =
    socket
    |> assign(:messages, messages)
   
  {:ok, socket}
end
```

Further down in the same live view, the `handle_info` function:

```elixir
def handle_info({{Message, :updated}, %{id: id, organisation_id: organisation_id}}, socket) do
  IO.puts("message id: #{id} for organisation_id: #{organisation_id} is for us!")

  messages =
    Messaging.query_messages_by_user(
      socket.assigns.organisation_id,
      socket.assigns.user_external_id
    )
    |> Repo.all()

  socket =
    socket
    |> assign(:messages, messages)

  {:noreply, socket}
end
```

I then modify a record using datagrip

Console output:

```bash
message id: e34c825d-52c4-4901-b2a9-94e24dc845b9 for organisation_id: 9a84a8b5-976c-4a64-82fd-8393e9fefe7f is for us!
message id: e34c825d-52c4-4901-b2a9-94e24dc845b9 for organisation_id: 9a84a8b5-976c-4a64-82fd-8393e9fefe7f is for us!
message id: e34c825d-52c4-4901-b2a9-94e24dc845b9 for organisation_id: 9a84a8b5-976c-4a64-82fd-8393e9fefe7f is for us!
message id: e34c825d-52c4-4901-b2a9-94e24dc845b9 for organisation_id: 9a84a8b5-976c-4a64-82fd-8393e9fefe7f is for us!
message id: e34c825d-52c4-4901-b2a9-94e24dc845b9 for organisation_id: 9a84a8b5-976c-4a64-82fd-8393e9fefe7f is for us!
message id: e34c825d-52c4-4901-b2a9-94e24dc845b9 for organisation_id: 9a84a8b5-976c-4a64-82fd-8393e9fefe7f is for us!
message id: e34c825d-52c4-4901-b2a9-94e24dc845b9 for organisation_id: 9a84a8b5-976c-4a64-82fd-8393e9fefe7f is for us!
message id: e34c825d-52c4-4901-b2a9-94e24dc845b9 for organisation_id: 9a84a8b5-976c-4a64-82fd-8393e9fefe7f is for us!
message id: e34c825d-52c4-4901-b2a9-94e24dc845b9 for organisation_id: 9a84a8b5-976c-4a64-82fd-8393e9fefe7f is for us!
message id: e34c825d-52c4-4901-b2a9-94e24dc845b9 for organisation_id: 9a84a8b5-976c-4a64-82fd-8393e9fefe7f is for us!
```

Oddly, the number of times the `handle_info` fires varies, a subsequent modification to another record that belongs to the same `organisation_id` only triggered 2 executions.

Have I incorrectly set things up perhaps?

Thanks again!

**Comment by cheerfulstoic** [comment]
Thanks very much!  I really enjoyed the discussion on Thinking Elixir! 😄 

First step I would try is to also do some `IO.puts`'s with your subscribes.  And both outputs should have the PID of the LiveView process so that you can see if you have multiple LiveView processes.  Also, for a bit more info I'd probably have an output outside of the `connected?` block just to understand what's going on.  So maybe something like this:

```elixir
def mount(params, _session, socket) do
  IO.puts("Before connected? check: #{inspect(self())}")
  if connected?(socket) do
    IO.puts("Inside connected? check: #{inspect(self())}")
    # EctoWatch.subscribe({Message, :inserted})
    # EctoWatch.subscribe({Message, :deleted})
   EctoWatch.subscribe({Message, :updated}, {:organisation_id, @org_id})
  end

  # ...

  {:ok, socket}
end

def handle_info({{Message, :updated}, %{id: id, organisation_id: organisation_id}}, socket) do
  IO.puts("message id: #{id} for organisation_id: #{organisation_id} is for us! #{inspect(self)}")

  # ...
```

If those are all showing the same PID, then there's probably something more to investigate with EctoWatch.

If they show different PIDs then my first guess would be that you have some browser tabs open to the LiveView page that you forgot about (happens to me all the time 😅 )

**Comment by camstuart** [comment]
Thanks @cheerfulstoic ! yeah I wondered if I had an extra browser tab stashed away, but I don't. I added the extra debugs as you suggested, and indeed there is another process in there somewhere. 

My live view that is doing this subscription actually passes the data down into a live component, I have a suspicion that's what is creating the extra pid, some type of re-render(mount) shenanigans.

I'm going to look at refactoring, and maybe move the EctoWatch subscription into the live component, or some other arrangement. The challenge I have is that parent needs access to the data (well, and count of records at least).

So I will experiment with different approaches and see if I can optimize! 

But as you say, `EctoWatch` is def doing it's job!

**Comment by cheerfulstoic** [comment]
👍  Good to know, thanks!

IIRC live components are run inside the same BEAM process as their parent.  So if a live component does a subscribe, it's all going to be the same process.  This is why you need the `phx-target={@myself}` to make the component handle an event.  But for just plain BEAM messages then (I think) you can't implement a `handle_info` in the component, just in the parent.  Then I think the general idea is that the parent is supposed to update it's state based on the messages so that the children can be updated (if their state changed)

I've been thinking for a little while that `EctoWatch` should have a `debug` option to output logs to help with this sort of thing.  The one thing I'm not 100% sure about how it will work is when doing the Phoenix PubSub broadcast then it's out of EctoWatch's hands because the `handle_info` is the next step.  But it might be possible to introspect the PubSub topic to figure out which process would be sent to 🤔   I also wonder, given [OTP 27's new process labels](https://www.elixirstreams.com/tips/process-labels) if those could be used to provide even better logs (e.g. to indicate when a message is being sent to a LiveView process vs. something else)

Thanks for the feedback.  Let me know what you find!

**Comment by cheerfulstoic** [comment]
FYI I just released version 0.11.0 with a `debug?` option.  You can learn more from [this HexDocs page](https://hexdocs.pm/ecto_watch/debugging.html)
